### PR TITLE
Mark Videos As Favorites

### LINKED ISSUE #684: Feature Request: Mark Video as Favorite
- Add button to mark/remove video as favorite
- List my favorite videos (filter on video list)
- Optionally include how many favorites a video has

### PR DESCRIPTION
**Pull request by KyleMcMaster**
- adds ability to toggle videos as favorites 
- filter video list by favorites
- adds `MemberFavoriteArchiveVideo` value object

fixes #684

### CODE REVIEW

File: src/DevBetterWeb.Core/Entities/ArchiveVideo.cs [hunk 1]
```diff
@@ -20,6 +21,7 @@ public class ArchiveVideo : BaseEntity, IAggregateRoot
   public int Views { get; set; } = 0;
 
   public List<Question> Questions { get; private set; } = new List<Question>();
+  public List<MemberFavoriteArchiveVideo> MemberFavorites { get; private set; } = new List<MemberFavoriteArchiveVideo>();
```
Comment (ardalis) [inline src/DevBetterWeb.Core/Entities/ArchiveVideo.cs:24]: You're probably following the bad examples already in this code, but generally in entities/aggregates you want to encapsulate collection properties, like this:

https://github.com/ardalis/CleanArchitecture/blob/main/src/Clean.Architecture.Core/ProjectAggregate/Project.cs#L12-L13

File: src/DevBetterWeb.Core/Entities/Member.cs [hunk 2]
```diff
@@ -66,7 +66,8 @@ internal Member(string userId, string firstName, string lastName)
   public decimal? CityLatitude { get; set; }
   public decimal? CityLongitude { get; set; }
   public List<BillingActivity> BillingActivities { get; set; } = new List<BillingActivity>();
-  //public List<MemberVideoProgress> Videos { get; private set; } = new List<MemberVideoProgress>();
+  public List<MemberFavoriteArchiveVideo> FavoriteArchiveVideos { get; set; } = new List<MemberFavoriteArchiveVideo>();
+  public List<MemberVideoProgress> Videos { get; private set; } = new List<MemberVideoProgress>();
```
Comment (ardalis) [inline src/DevBetterWeb.Core/Entities/Member.cs:70]: Same feedback here regarding encapsulation.

Also if we're using the aggregate design pattern (which we don't seem to be but probably should be) then you'd want to pick from one side or the other to be the "owner" of favorites. That is, either a member adds a favorite video, or a video adds a favoriting member, but not both. This help ensure there is only one place where you need to perform validation on this operation, among other things.

File: src/DevBetterWeb.Core/Entities/Member.cs [hunk 3]
```diff
@@ -119,6 +120,26 @@ public void UpdateAddress(string? address)
...
+
+  public void RemoveFavoriteArchiveVideo(ArchiveVideo archiveVideo)
+  {
+    if (FavoriteArchiveVideos.Any(fav => fav.ArchiveVideoId == archiveVideo.Id))
+    {
+	  var removal = FavoriteArchiveVideos.First(v => v.ArchiveVideoId == archiveVideo.Id);
+
+  	  FavoriteArchiveVideos.Remove(removal);
+    }
+  }
```
Comment (ardalis) [inline src/DevBetterWeb.Core/Entities/Member.cs:141]: Looks good.

**Review by ardalis — CHANGES_REQUESTED** [review]
Looks pretty good! Remind me I need to make db changes as part of this deployment.

Thanks!
Steve

File: src/DevBetterWeb.Core/Entities/Member.cs [hunk 4]
```diff
@@ -66,7 +66,8 @@ internal Member(string userId, string firstName, string lastName)
   public decimal? CityLatitude { get; set; }
   public decimal? CityLongitude { get; set; }
   public List<BillingActivity> BillingActivities { get; set; } = new List<BillingActivity>();
-  //public List<MemberVideoProgress> Videos { get; private set; } = new List<MemberVideoProgress>();
```
Comment (KyleMcMaster) [inline src/DevBetterWeb.Core/Entities/Member.cs:69]: This shouldn't have been removed

**Pull request merged**